packages/terra-functional-testing/lib/services/wdio-selenium-docker-service.js

```diff
@@ -0,0 +1,164 @@
+/* eslint-disable class-methods-use-this */
+const path = require('path');
+const util = require('util');
+const childProcess = require('child_process');
+const Logger = require('../logger/logger');
+
+const exec = util.promisify(childProcess.exec);
+const logger = new Logger({ prefix: 'wdio-selenium-docker-service' });
+
+class SeleniumDockerService {
+  /**
+   * Prepares the docker testing environment.
+   */
+  async onPrepare(config) {
+    this.host = config.hostname;
+    this.port = config.port;
+
+    await this.initializeSwarm();
+    await this.deployStack();
+  }
+
+  /**
+   * Initializes a docker swarm instance.
+   */
+  async initializeSwarm() {
+    const { stdout: dockerInfo } = await exec('docker info --format "{{json .}}"');
+    const { Swarm } = JSON.parse(dockerInfo);
+
+    if (Swarm.LocalNodeState !== 'active') {
+      logger.log('Initializing docker swarm.');
+
+      await exec('docker swarm init');
+    }
+  }
+
+  /**
+   * Deploys the docker stack. The previous stack will be removed if it exists.
+   */
+  async deployStack() {
+    // Remove the previous stack if one exists.
+    await this.removeStack();
+
+    logger.log('Deploying docker stack.');
+
+    const composeFilePath = path.resolve(__dirname, '../docker/docker-compose.yml');
+
+    await exec(`docker stack deploy -c ${composeFilePath} wdio`);
+
+    await this.waitForNetworkReady();
+  }
+
+  /**
+   * Removes the docker stack.
+   */
+  async removeStack() {
+    const { stdout: stackInfo } = await exec('docker stack ls | grep wdio || true');
+
+    if (stackInfo) {
+      logger.log('Removing docker stack.');
+
+      await exec('docker stack rm wdio');
+
+      // Ensure the services and network have been removed.
+      await this.waitForServiceRemoval();
+      await this.waitForNetworkRemoval();
+    }
+  }
+
+  /**
+   * Waits for a command to complete successfully.
+   * @param {string} command - The shell command to run.
+   * @param {func} callback - A callback function to accept or reject the result of the command. Must return a promise.
+   * @param {number} retries - The number of times to retry the command. Defaults to 30 retries.
+   * @param {number} interval - The timeout between commands in milliseconds. Defaults to every two seconds.
+   * @returns {Promise} - A promise that resolves when the callback accepts the command response.
+   */
+  async pollCommand(command, callback, retries = 30, interval = 2000) {
```

Comment:
curious, how did we land on 30 retries?

Reply:
It's an arbitrary value intended to give these command attempts 60 seconds to start/stop. 

Interval - Every two seconds
Duration - Timeout after 1 minute

2000ms * 30ms = 60000ms = 1 minute